README.md
# A simple usage example for Java

## Java Usage 

The following example shows a server and a client using the `com.googlecode.protobuf.socketrpc` Java API.

Note that this example uses delimited rpc channel factories which is new in version 2.0 and is not compatible with 1.x versions. Use undelimited rpc channel factories if you need to communicate with 1.x versions.

### Server side

```
// Start server
ServerRpcConnectionFactory rpcConnectionFactory = SocketRpcConnectionFactories
    .createServerRpcConnectionFactory(port);
RpcServer server = new RpcServer(rpcConnectionFactory, 
    Executors.newFixedThreadPool(threadPoolSize), true);
server.registerService(new MyServiceImpl()); // For non-blocking impl
server.registerBlockingService(MyService
    .newReflectiveBlockingService(new MyBlockingInterfaceImpl())); // For blocking impl
server.run();
```

### Client Side (Blocking)

```
// Create channel
RpcConnectionFactory connectionFactory = SocketRpcConnectionFactories
    .createRpcConnectionFactory(host, port);
BlockingRpcChannel channel = RpcChannels.newBlockingRpcChannel(connectionFactory);

// Call service
BlockingInterface service = MyService.newBlockingStub(channel);
RpcController controller = new SocketRpcController();
MyResponse myResponse = service.myMethod(controller, myRequest);

// Check success
if (rpcController.failed()) {
  System.err.println(String.format("Rpc failed %s : %s", rpcController.errorReason(),
      rpcController.errorText()));
}
```

### Client Side (non-blocking)

```
// Create a thread pool
ExecutorService threadPool = Executor.newFixedThreadPool(1);

// Create channel
RpcConnectionFactory connectionFactory = SocketRpcConnectionFactories
    .createRpcConnectionFactory(host, port);
RpcChannel channel = RpcChannels.newRpcChannel(connectionFactory, threadPool);

// Call service
MyService myService = MyService.newStub(channel);
RpcController controller = new SocketRpcController();
myService.myMethod(rpcController, myRequest,
    new RpcCallback<MyResponse>() {
      public void run(MyResponse myResponse) {
        System.out.println("Received Response: " + myResponse);
      }
    });
    
// Check success
if (rpcController.failed()) {
  System.err.println(String.format("Rpc failed %s : %s", rpcController.errorReason(),
      rpcController.errorText()));
}
```

## Python Usage

The following example shows a server and a client using the `protobuf.socketrpc` Python API.

### Server side

```
// Start server
server = protobuf.socketrpc.server.SocketRpcServer(port)
server.registerService(MyServiceImpl())
server.run()
```

### Client Side

```
// Define callback
class Callback:
  def run(self,response):
    print "Received Response: %s" % response

// Create channel
channel = protobuf.socketrpc.channel.SocketRpcChannel(hostname,port)
controller = channel.newController()

// Call service
service  = MyService_Stub(channel)
service.myMethod(controller,request,Callback())

// Check success
if controller.failed():
  print "Rpc failed %s : %s" % (controller.error,controller.reason)
```

### Further examples

There are two more examples included in the source tarball. You might also want to have a look at the unit tests.


python/ChangeLog
Version 1.3.x (SVN Trunk) YYYY-MM-DD

* Protobuf 2.4.0 compatible

Version 1.3.2 (SVN Revision 89) 2010-11-04

* Use a more pythonic directory structure to allow better use of
  setuptools/distribute features
* Add metadata to setup.py
* Rename Python module from protobuf to protobuf.socketrpc to avoid
  name clashes
* Join thread instead of busy waiting for synchronous calls
* Protobuf 2.3.0 compatible
* Fixes for Issue 11 (Cannot create different server side handlers for
  the same service) and Issue 12 (Exceptions on server not propogated
  to client)
* Fixed unit tests

Version 1.3.1 (SVN Revision 46) 2009-12-21

* Added Service class to make rpc calls cleaner and hide some of the
  setup details

Version 1.3 (SVN Revision 33) 2009-10-04

* Protobuf 2.2.0 compatible and dependent.
* Moving error reasons out of response message into its own top level
  type in the proto. Added client side error reasons to the error
  codes to be consistent.

Version 1.2 (SVN Revision 12) 2009-02-14

* Errors that occur on server side are not relayed back to
  client. Client has new error reasons for them in the controller.
* Client callback behavior is faithful replication of server callback
  scenarios:
  - Callback is called with good response.
  - Callback is not called.
  - Callback is called with null.
  - Callback is called with empty protocol buffer.
* Added comprehensive unit tests for client and server.

Version 1.1 (SVN Revision 9) 2009-02-06

* Added error handling. When errors occur, the client can find out
  what error occurred from the controller.
* Started adding unit tests.


python/src/protobuf/socketrpc/rpc_pb2.py
#!/usr/bin/python
# Copyright (c) 2011 Las Cumbres Observatory (www.lcogt.net)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
# Generated by the protocol buffer compiler.  DO NOT EDIT!

from google.protobuf import descriptor
from google.protobuf import message
from google.protobuf import reflection
from google.protobuf import descriptor_pb2
# @@protoc_insertion_point(imports)



DESCRIPTOR = descriptor.FileDescriptor(
  name='rpc.proto',
  package='protobuf.socketrpc',
  serialized_pb='\n\trpc.proto\x12\x12protobuf.socketrpc\"K\n\x07Request\x12\x14\n\x0cservice_name\x18\x01 \x02(\t\x12\x13\n\x0bmethod_name\x18\x02 \x02(\t\x12\x15\n\rrequest_proto\x18\x03 \x02(\x0c\"\x81\x01\n\x08Response\x12\x16\n\x0eresponse_proto\x18\x01 \x01(\x0c\x12\r\n\x05\x65rror\x18\x02 \x01(\t\x12\x17\n\x08\x63\x61llback\x18\x03 \x01(\x08:\x05\x66\x61lse\x12\x35\n\x0c\x65rror_reason\x18\x04 \x01(\x0e\x32\x1f.protobuf.socketrpc.ErrorReason*\xd9\x01\n\x0b\x45rrorReason\x12\x14\n\x10\x42\x41\x44_REQUEST_DATA\x10\x00\x12\x15\n\x11\x42\x41\x44_REQUEST_PROTO\x10\x01\x12\x15\n\x11SERVICE_NOT_FOUND\x10\x02\x12\x14\n\x10METHOD_NOT_FOUND\x10\x03\x12\r\n\tRPC_ERROR\x10\x04\x12\x0e\n\nRPC_FAILED\x10\x05\x12\x19\n\x15INVALID_REQUEST_PROTO\x10\x06\x12\x16\n\x12\x42\x41\x44_RESPONSE_PROTO\x10\x07\x12\x10\n\x0cUNKNOWN_HOST\x10\x08\x12\x0c\n\x08IO_ERROR\x10\tB4\n!com.googlecode.protobuf.socketrpcB\x0fSocketRpcProtos')

_ERRORREASON = descriptor.EnumDescriptor(
  name='ErrorReason',
  full_name='protobuf.socketrpc.ErrorReason',
  filename=None,
  file=DESCRIPTOR,
  values=[
    descriptor.EnumValueDescriptor(
      name='BAD_REQUEST_DATA', index=0, number=0,
      options=None,
      type=None),
    descriptor.EnumValueDescriptor(
      name='BAD_REQUEST_PROTO', index=1, number=1,
      options=None,
      type=None),
    descriptor.EnumValueDescriptor(
      name='SERVICE_NOT_FOUND', index=2, number=2,
      options=None,
      type=None),
    descriptor.EnumValueDescriptor(
      name='METHOD_NOT_FOUND', index=3, number=3,
      options=None,
      type=None),
    descriptor.EnumValueDescriptor(
      name='RPC_ERROR', index=4, number=4,
      options=None,
      type=None),
    descriptor.EnumValueDescriptor(
      name='RPC_FAILED', index=5, number=5,
      options=None,
      type=None),
    descriptor.EnumValueDescriptor(
      name='INVALID_REQUEST_PROTO', index=6, number=6,
      options=None,
      type=None),
    descriptor.EnumValueDescriptor(
      name='BAD_RESPONSE_PROTO', index=7, number=7,
      options=None,
      type=None),
    descriptor.EnumValueDescriptor(
      name='UNKNOWN_HOST', index=8, number=8,
      options=None,
      type=None),
    descriptor.EnumValueDescriptor(
      name='IO_ERROR', index=9, number=9,
      options=None,
      type=None),
  ],
  containing_type=None,
  options=None,
  serialized_start=243,
  serialized_end=460,
)


BAD_REQUEST_DATA = 0
BAD_REQUEST_PROTO = 1
SERVICE_NOT_FOUND = 2
METHOD_NOT_FOUND = 3
RPC_ERROR = 4
RPC_FAILED = 5
INVALID_REQUEST_PROTO = 6
BAD_RESPONSE_PROTO = 7
UNKNOWN_HOST = 8
IO_ERROR = 9



_REQUEST = descriptor.Descriptor(
  name='Request',
  full_name='protobuf.socketrpc.Request',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    descriptor.FieldDescriptor(
      name='service_name', full_name='protobuf.socketrpc.Request.service_name', index=0,
      number=1, type=9, cpp_type=9, label=2,
      has_default_value=False, default_value=unicode("", "utf-8"),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      options=None),
    descriptor.FieldDescriptor(
      name='method_name', full_name='protobuf.socketrpc.Request.method_name', index=1,
      number=2, type=9, cpp_type=9, label=2,
      has_default_value=False, default_value=unicode("", "utf-8"),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      options=None),
    descriptor.FieldDescriptor(
      name='request_proto', full_name='protobuf.socketrpc.Request.request_proto', index=2,
      number=3, type=12, cpp_type=9, label=2,
      has_default_value=False, default_value="",
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      options=None),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  options=None,
  is_extendable=False,
  extension_ranges=[],
  serialized_start=33,
  serialized_end=108,
)


_RESPONSE = descriptor.Descriptor(
  name='Response',
  full_name='protobuf.socketrpc.Response',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    descriptor.FieldDescriptor(
      name='response_proto', full_name='protobuf.socketrpc.Response.response_proto', index=0,
      number=1, type=12, cpp_type=9, label=1,
      has_default_value=False, default_value="",
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      options=None),
    descriptor.FieldDescriptor(
      name='error', full_name='protobuf.socketrpc.Response.error', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=unicode("", "utf-8"),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      options=None),
    descriptor.FieldDescriptor(
      name='callback', full_name='protobuf.socketrpc.Response.callback', index=2,
      number=3, type=8, cpp_type=7, label=1,
      has_default_value=True, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      options=None),
    descriptor.FieldDescriptor(
      name='error_reason', full_name='protobuf.socketrpc.Response.error_reason', index=3,
      number=4, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      options=None),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  options=None,
  is_extendable=False,
  extension_ranges=[],
  serialized_start=111,
  serialized_end=240,
)

_RESPONSE.fields_by_name['error_reason'].enum_type = _ERRORREASON
DESCRIPTOR.message_types_by_name['Request'] = _REQUEST
DESCRIPTOR.message_types_by_name['Response'] = _RESPONSE

class Request(message.Message):
  __metaclass__ = reflection.GeneratedProtocolMessageType
  DESCRIPTOR = _REQUEST
  
  # @@protoc_insertion_point(class_scope:protobuf.socketrpc.Request)

class Response(message.Message):
  __metaclass__ = reflection.GeneratedProtocolMessageType
  DESCRIPTOR = _RESPONSE
  
  # @@protoc_insertion_point(class_scope:protobuf.socketrpc.Response)

# @@protoc_insertion_point(module_scope)
